### PR TITLE
Build artifact with sources with submodules in Darwin special build

### PR DESCRIPTION
**Pull request by qoega**
### Changelog category (leave one):
- Build/Testing/Packaging Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Resource with source code including submodules is built in Darwin special build task. It may be used to build ClickHouse without checkouting submodules.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
This is an automated comment for commit f85b88b5b8fc9dae5b4f4bdcc60102634726366c with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/51435/f85b88b5b8fc9dae5b4f4bdcc60102634726366c/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: docker/packager/binary/build.sh [hunk 1]
```diff
@@ -15,6 +15,11 @@ if [ "$EXTRACT_TOOLCHAIN_DARWIN" = "1" ]; then
   mkdir -p /build/cmake/toolchain/darwin-x86_64
   tar xJf /MacOSX11.0.sdk.tar.xz -C /build/cmake/toolchain/darwin-x86_64 --strip-components=1
   ln -sf darwin-x86_64 /build/cmake/toolchain/darwin-aarch64
+
+  if [ "$EXPORT_SOURCES_WITH_SUBMODULES" = "1" ]; then
+    cd /build
+    tar --exclude-vcs-ignores --exclude-vcs --exclude build --exclude build_docker --exclude debian --exclude .git --exclude .github --exclude .cache --exclude docs --exclude tests/integration -c . | pigz -9 > /output/source_sub.tar.gz
```
Comment (Felixoid) [inline docker/packager/binary/build.sh:21]: what about using xz/zstd?

File: docker/packager/packager [hunk 2]
```diff
@@ -168,6 +168,7 @@ def parse_env_variables(
             "-DCMAKE_TOOLCHAIN_FILE=/build/cmake/darwin/toolchain-x86_64.cmake"
         )
         result.append("EXTRACT_TOOLCHAIN_DARWIN=1")
+        result.append("EXPORT_SOURCES_WITH_SUBMODULES=1")
```
Comment (Felixoid) [inline docker/packager/packager:171]: so, it will be always in the `clickhouse-builds`, but only for darwin build. Looks reasonable

**Comment by Felixoid** [comment]
To reduce the size of the archive, we can additionally add this part to the workflow OR to the script https://github.com/ClickHouse/ClickHouse/blob/master/.github/workflows/pull_request.yml#L558-L564

File: docker/packager/binary/build.sh (hunk 1 above)
Comment (qoega) [reply docker/packager/binary/build.sh:21]: GitHub provides tar.gz there. I wanted minimal changed compared to official way to get sources

File: docker/packager/packager (hunk 2 above)
Comment (qoega) [reply docker/packager/packager:171]: Only for one of 2 darwin builds. It was pretty fast in CI.

**Comment by qoega** [comment]
> To reduce the size of the archive, we can additionally add this part to the workflow OR to the script https://github.com/ClickHouse/ClickHouse/blob/master/.github/workflows/pull_request.yml#L558-L564

It should be unnecessary. I try to fix tar `--exclude` filters and not include any .git files or similar. Only one version so this change is unnecessary to minimise tar size.

**Comment by Felixoid** [comment]
```
> ls /tmp/source_sub* -l
-rw-r--r-- 1 felixoid felixoid 380849752 Jul  3 16:47 /tmp/source_sub_sparsed.tar.gz
-rw-r--r-- 1 felixoid felixoid 283703728 Jul  3 17:01 /tmp/source_sub_sparsed.tar.xz
-rw-r--r-- 1 felixoid felixoid 370023959 Jul  3 17:02 /tmp/source_sub_sparsed.tar.zst
-rw-r--r-- 1 felixoid felixoid 791041769 Jul  3 15:37 /tmp/source_sub.tar.gz
-rw-r--r-- 1 felixoid felixoid 508762888 Jul  3 16:01 /tmp/source_sub.tar.xz
-rw-r--r-- 1 felixoid felixoid 715459283 Jul  3 16:43 /tmp/source_sub.tar.zst
```

Sparse checkout is exactly two times lighter. We should use it.

`xz` is a quite long on my laptop. I think, `gz` looks OK here

**Comment by qoega** [comment]
>Sparse checkout is exactly two times lighter. We should use it.

OK. I just don't understand where I should add this step. If you can patch this PR the way you want - you are welcome.

**Comment by Felixoid** [comment]
Than should do the thing

File: docker/packager/packager (hunk 2 above)
Comment (Felixoid) [reply docker/packager/packager:171]: Just found, that we don't have the same thing for `is_cross_darwin_arm`

**Comment by Felixoid** [comment]
467 vs 870. Looks worth!

**Comment by Felixoid** [comment]
Let's do it for both darwin builds and we're good to go!

**Comment by qoega** [comment]
One should be enough

File: docker/packager/packager (hunk 2 above)
Comment (Felixoid) [reply docker/packager/packager:171]: Got it, the artifact will be the same, so we need only one of them

**Review by Felixoid — APPROVED** [review]
we're good to go after the green checks

**Pull request merged**